fedmsg/crypto/utils.py

```diff
@@ -7,6 +7,52 @@
 _log = logging.getLogger(__name__)
 
 
+def fix_datagrepper_message(message):
+    """
+    See if a message is (probably) a datagrepper message and attempt to mutate
+    it to pass signature validation.
+
+    Datagrepper adds the 'source_name' and 'source_version' keys. If messages happen
+    to use those keys, they will fail message validation. Additionally, a 'headers'
+    dictionary is present on all responses, regardless of whether it was in the
+    original message or not. This is deleted if it's null, which won't be correct in
+    all cases. Finally, datagrepper turns the 'timestamp' field into a float, but it
+    might have been an integer when the message was signed.
+
+    A copy of the dictionary is made and returned if altering the message is necessary.
+
+    I'm so sorry.
+
+    Args:
+        message (dict): A message to clean up.
+
+    Returns:
+        dict: A copy of the provided message, with the datagrepper-related keys removed
+            if they were present.
+    """
+    if not ('source_name' in message and 'source_version' in message):
+        return message
+
+    # Don't mutate the original message
+    message = message.copy()
+
+    del message['source_name']
+    del message['source_version']
+    # datanommer adds the headers field to the message in all cases.
+    # This is a huge problem because if the signature was generated with a 'headers'
+    # key set and we delete it here, messages will fail validation, but if we don't
+    # messages will fail validation if they didn't have a 'headers' key set.
```

Comment:
So here's another crazy idea: if headers is null and validation fals, should we try adding a null headers back and see if *that* version passes validation?

Reply:
We could, and I considered that. The reason I opted not to is because this is already way more complicated than it should be and even if we did that we'd still be wrong plenty of times. What happens when someone sets that to ``null`` rather than an empty object? What happens when messages get signed with floating point timestamps rather than ints? datanommer truncates those. What happens when datanommer adds a new field and forgets to adjust this?

I'd rather have datanommer make strong promises about not altering messages.